rust/lance/src/index/vector/ivf/shuffler.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Disk-based shuffle a stream of [RecordBatch] into each IVF partition.
+//!
+//! 1. write the entire stream to a file
+//! 2. count the number of rows in each partition
+//! 3. read the data back into memory and shuffle into grouped vectors
+//!
+//! Problems for the future:
+//! 1. while groupby column will stay the same, we may want to include extra data columns in the future
+//! 2. shuffling into memory is fast but we should add disk buffer to support bigger datasets
+
+use arrow_array::cast::{as_fixed_size_list_array, as_primitive_array};
+use arrow_array::types::{UInt32Type, UInt64Type, UInt8Type};
+use arrow_array::UInt32Array;
+use futures::{stream, StreamExt};
+use kdam::BarExt;
+use lance_core::datatypes::Schema;
+use lance_core::io::{FileReader, FileWriter, ReadBatchParams, RecordBatchStream};
+
+use lance_core::io::object_store::ObjectStore;
+use lance_core::{Error, Result, ROW_ID};
+use lance_index::vector::{PART_ID_COLUMN, PQ_CODE_COLUMN};
+use log::info;
+use object_store::path::Path;
+use snafu::{location, Location};
+use tempfile::TempDir;
+
+const UNSORTED_BUFFER: &str = "unsorted.lance";
+
+fn get_temp_dir() -> Result<Path> {
+    let dir = TempDir::new()?;
+    let tmp_dir_path = Path::from_filesystem_path(dir.path()).map_err(|e| Error::IO {
+        message: format!("failed to get buffer path in shuffler: {}", e),
+        location: location!(),
+    })?;
+    Ok(tmp_dir_path)
+}
+
+pub struct IvfShuffler {
+    pub num_partitions: u32,
```

Comment:
Are these attribute needs to be public?

Reply:
nope, these are left over from test, will fix